pkg/services/mto_service_item/mto_service_item_creator.go

```
@@ -163,6 +202,54 @@ func (o *mtoServiceItemCreator) CreateMTOServiceItem(serviceItem *models.MTOServ
 	return &createdServiceItems, nil, nil
 }
 
+// checkDuplicateServiceCodes checks if the move or shipment has a duplicate service item with the same code as the one
+// requested.
+func (o *mtoServiceItemCreator) checkDuplicateServiceCodes(serviceItem *models.MTOServiceItem) error {
+	var duplicateServiceItem models.MTOServiceItem
+
+	queryFilters := []services.QueryFilter{
+		query.NewQueryFilter("move_id", "=", serviceItem.MoveTaskOrderID),
+		query.NewQueryFilter("re_service_id", "=", serviceItem.ReServiceID),
+	}
+	if serviceItem.MTOShipmentID != nil {
+		queryFilters = append(queryFilters, query.NewQueryFilter("mto_shipment_id", "=", serviceItem.MTOShipmentID))
+	}
+
+	err := o.builder.FetchOne(&duplicateServiceItem, queryFilters)
+	if err == nil && duplicateServiceItem.ID != uuid.Nil {
```

What happens if there is an err? we should return InternalServerError and rollback I would think.

---

If there is an error, it means we didn't find the item which is good. I don't think this function silently fails if the record isn't found, but I'll check to make sure I'm doing the right thing (maybe I should just remove the uuid check?)

---

or if we don't need the err for anything, you can use the builder standalone and then perform the check on the duplicate item id. could the err ever be anything other than a not found?

---

it could be! I went ahead and added a clause after this that specifically checks if it's a not found error, and returns the err if not

---

(it has to check the error message because there's no good typing here to help me out)